rocketpy/motors/Tank.py

```diff
@@ -0,0 +1,842 @@
+        self.upperCapLiquidCentroid = baseHeight + self.upperCap.filled_centroid
+        self.upperCapGasCentroid = baseHeight + self.upperCap.empty_centroid
+        self.upperCapCentroid = baseHeight + self.upperCap.centroid
+
+    def evaluateRelativeDistances(self, t):
+        """Calculates the relative distances of the centroids of liquid and gaseous
+        tank portions to its center of mass.
+
+        Parameters
+        ----------
+        t : float
+            Time at which the relative distances are to be calculated.
+
+        Returns
+        -------
+        None
+        """
+        self.bottomCapRelDistLiq = self.centerOfMass(t) - self.bottomCapLiquidCentroid
+        self.bottomCapRelDistGas = self.centerOfMass(t) - self.bottomCapGasCentroid
+        self.cylinderRelDistLiq = self.centerOfMass(t) - self.cylinderLiquidCentroid
+        self.cylinderRelDistGas = self.centerOfMass(t) - self.cylinderGasCentroid
+        self.upperCapRelDistLiq = self.centerOfMass(t) - self.upperCapLiquidCentroid
+        self.upperCapRelDistGas = self.centerOfMass(t) - self.upperCapGasCentroid
+
+    @functools.cached_property
+    def volume(self):
+        """Returns the total volume of the tank structure.
+
+        Returns
+        -------
+        float
+            Tank's total volume.
+        """
+        return self.bottomCap.volume + self.cylinder.volume + self.upperCap.volume
+
+    @abstractmethod
+    def mass(self):
+        """Returns the total mass of liquid and gases inside the tank as a
+        function of time.
+
+        Parameters
+        ----------
+        time : float
+            Time in seconds.
+
+        Returns
+        -------
+        Function
+            Mass of the tank as a function of time. Units in kg.
+        """
+        pass
+
+    @abstractmethod
+    def netMassFlowRate(self):
+        """Returns the net mass flow rate of the tank as a function of time.
+        Net mass flow rate is the mass flow rate entering the tank minus the
+        mass flow rate exiting the tank, including liquids and gases. Positive
+        is defined as a net mass flow rate entering the tank.
+
+        Parameters
+        ----------
+        time : float
+            Time in seconds.
+
+        Returns
+        -------
+        Function
+            Net mass flow rate of the tank as a function of time.
+            Positive is defined as a net mass flow rate entering the tank.
+        """
+        pass
+
+    @abstractmethod
+    def liquidVolume(self):
+        """Returns the volume of liquid inside the tank as a function
+        of time.
+
+        Parameters
+        ----------
+        time : float
+            Time in seconds.
+
+        Returns
+        -------
+        Function
+            Tank's liquid volume as a function of time.
+        """
+        pass
+
+    @functools.cached_property
+    def centerOfMass(self):
+        """Returns the center of mass of the tank's fluids as a function of
+        time.
```

Comment:
TODO: Specify relative to which point is the center of mass calculated (coordinate system orientation and origin).